src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/ConditionalWeakTable.cs

```diff
@@ -523,10 +523,20 @@ internal int FindEntry(TKey key, out object? value)
                 int bucket = hashCode & (_buckets.Length - 1);
                 for (int entriesIndex = Volatile.Read(ref _buckets[bucket]); entriesIndex != -1; entriesIndex = _entries[entriesIndex].Next)
                 {
-                    if (_entries[entriesIndex].HashCode == hashCode && _entries[entriesIndex].depHnd.GetPrimaryAndSecondary(out value) == key)
+                    if (_entries[entriesIndex].HashCode == hashCode)
                     {
-                        GC.KeepAlive(this); // ensure we don't get finalized while accessing DependentHandles.
-                        return entriesIndex;
+                        DependentHandle dependentHandle = _entries[entriesIndex].depHnd;
+                        object? target = dependentHandle.UnsafeGetTarget();
+
+                        if (target == key)
+                        {
+                            value = dependentHandle.UnsafeGetDependent();
```

Comment:
Just want to point out: This change is actually fixing a potential GC hole.

The `GetPrimaryAndSecondary` FCalls assumed that the `out` argument is on the stack. It is not guaranteed since it is passed all the way from user code. If the user code passed a location on GCHeap for the `ConditionalWeakTable.TryGetValue` `out` argument  (note that it can happen implicitly - for example, if the method is async), there will be GC write barrier missing that can lead to intermittent crashes or data corruptions.

Reply:
Oh wow, right! I was indeed a bit confused by the fact that that FCall was taking an `out object` and in the VM it was just `Object**` without pinning or anything, but I figured the runtime was just doing some magic marshalling somewhere to fix that (I'm not too familiar with the FCall implementation details). If that wasn't actually the case, glad to hear we've now spotted and fixed that, that's awesome! And thanks for the additional explanation! 😄